### PR TITLE
Add JobURLConfig option to Plank

### PR DESCRIPTION
**Pull request by alvaroaleman**
As we have some private and some public repos being tested via the same Prow instance, we need a way to configure `JobURLPrefix` on a per-repo basis or we will be stuck with `job_url_template` forever.

/assign @stevekuznetsov

### CODE REVIEW

File: prow/config/config.go [hunk 1]
```diff
@@ -176,7 +176,30 @@ type Plank struct {
 	DefaultDecorationConfig *prowapi.DecorationConfig `json:"default_decoration_config,omitempty"`
 	// JobURLPrefix is the host and path prefix under
 	// which job details will be viewable
+	// It should never be accessed directly, use `GetJobURLPrefix` instead
 	JobURLPrefix string `json:"job_url_prefix,omitempty"`
```
Comment (alvaroaleman) [inline prow/config/config.go]: I really hate that this is public, but unfortunately it is required for the YAML parsing of the config

File: prow/config/config_test.go [hunk 2]
```diff
@@ -1843,3 +1843,178 @@ github_reporter:
...
+					}}},
+			refs:                 &prowapi.Refs{Org: "my-alternate-org", Repo: "my-second-repo"},
+			expectedJobURLPrefix: "https://my-third-prow",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			if prefix := tc.plank.JobURLPrefix(tc.refs); prefix != tc.expectedJobURLPrefix {
+				t.Errorf("expected JobUrlPrefix to be '%s' but was '%s'", tc.expectedJobURLPrefix, prefix)
```
Comment (stevekuznetsov) [inline prow/config/config_test.go]: nit: `'%s'` is `%q`

File: prow/config/config_test.go [hunk 3]
```diff
@@ -1843,3 +1843,178 @@ github_reporter:
...
+			}
+		})
+	}
+}
+
+func TestValidateComponentConfig(t *testing.T) {
+	testCases := []struct {
+		Name       string
+		config     *Config
+		errEpexted bool
```
Comment (stevekuznetsov) [inline prow/config/config_test.go]: Typo

File: prow/config/config_test.go [hunk 4]
```diff
@@ -1843,3 +1843,178 @@ github_reporter:
...
+			if prefix := tc.plank.JobURLPrefix(tc.refs); prefix != tc.expectedJobURLPrefix {
+				t.Errorf("expected JobUrlPrefix to be '%s' but was '%s'", tc.expectedJobURLPrefix, prefix)
+			}
+		})
+	}
+}
+
+func TestValidateComponentConfig(t *testing.T) {
+	testCases := []struct {
+		Name       string
```
Comment (stevekuznetsov) [inline prow/config/config_test.go]: Why is this exported?

File: prow/spyglass/spyglass_test.go [hunk 5]
```diff
@@ -100,14 +100,14 @@ test/e2e/e2e.go:137 BeforeSuite on Node 1 failed test/e2e/e2e.go:137
 			BucketName: "test-bucket",
 			Name:       "logs/example-ci-run/403/started.json",
 			Content: []byte(`{
-						  "node": "gke-prow-default-pool-3c8994a8-qfhg",
```
Comment (stevekuznetsov) [inline prow/spyglass/spyglass_test.go]: Can we nix the whitespace changes here or pull them into a different commit?

**Comment by alvaroaleman** [comment]
> Maybe let's deprecate JobURLPrefix (if you literally add a //Deprecated comment editors will even grey it out) and instead of having an override, just have a mapping of prefixes. This is another place where we should if possible emulate branchprotector structure in config. Dang would be nice to have generics or create some sort of interface here for a structured set of config that can hold overrides.

Yeah, perfectly fine with me as well. 

>  instead of having an override, just have a mapping of prefixes

By that you mean basically the same thing I implemented except that there is no default?

File: prow/config/config_test.go (hunk 2 above)
Comment (alvaroaleman) [reply prow/config/config_test.go]: TILd

**Comment by stevekuznetsov** [comment]
> By that you mean basically the same thing I implemented except that there is no default?

Yeah, in other configs we have a `'*'` field that can be explicitly used as a default -- to make this backwards compat we should populate that field in the explicit mapping if a config sets the current prefix field, when we remove the deprecated field we will require that people explicitly set their default.

**Comment by cjwagner** [comment]
SGTM.
> Dang would be nice to have generics or create some sort of interface here for a structured set of config that can hold overrides.

ref: https://github.com/kubernetes/test-infra/issues/9757

**Comment by alvaroaleman** [comment]
@stevekuznetsov Updated, PTAL

**Comment by alvaroaleman** [comment]
> I think this looks good, we want to make the announcement in ANNOUNCEMENTS.md and add a future TODO to remove the other field

Should I add that to the PR?

**Comment by stevekuznetsov** [comment]
Yes please!

File: prow/config/config.go [hunk 6]
```diff
@@ -174,9 +174,32 @@ type Plank struct {
 	// DefaultDecorationConfig are defaults for shared fields for ProwJobs
 	// that request to have their PodSpecs decorated
 	DefaultDecorationConfig *prowapi.DecorationConfig `json:"default_decoration_config,omitempty"`
+	// Deprecated, use JobURLPrefixConfig instead
 	// JobURLPrefix is the host and path prefix under
 	// which job details will be viewable
+	// TODO @alvaroaleman: Remove in September 2019
```
Comment (alvaroaleman) [inline prow/config/config.go]: I'll create an appointment in my calendar once this merges

**Comment by stevekuznetsov** [comment]
Since this is user-facing let's give @cjwagner @fejta one last chance to look at this before we merge
/lgtm
/hold

@alvaroaleman thanks :)

File: prow/config/config.go [hunk 7]
```diff
@@ -174,9 +174,32 @@ type Plank struct {
...
 	JobURLPrefix string `json:"job_url_prefix,omitempty"`
+	// JobURLPrefixConfig is the host and path prefix under which job details
+	// will be viewable. Use `*` as key for your global default.
+	JobURLPrefixConfig map[string]JobURLPrefixConfig `json:"job_url_prefix_config,omitempty"`
```
Comment (cjwagner) [inline prow/config/config.go]: Please indicate what this map is keyed by in the comment. As is, this is not enough information for me to provide a valid configuration.

File: prow/config/config.go [hunk 8]
```diff
@@ -174,9 +174,32 @@ type Plank struct {
...
+	if refs == nil {
+		return p.JobURLPrefixConfig["*"].URL
+	}
+	if p.JobURLPrefixConfig[refs.Org].Repos[refs.Repo] != "" {
+		return p.JobURLPrefixConfig[refs.Org].Repos[refs.Repo]
+	}
+	if p.JobURLPrefixConfig[refs.Org].URL != "" {
+		return p.JobURLPrefixConfig[refs.Org].URL
+	}
+	return p.JobURLPrefixConfig["*"].URL
```
Comment (cjwagner) [inline prow/config/config.go]: Instead of using multiple levels of maps and defaults, I think it would be preferable to use the pattern that we use elsewhere (especially for consistency): use a `map[string]string` where the keys are either `org` or `org/repo` format and check for values in the following order `org/repo` > `org` > `*`.

Comment (stevekuznetsov) [reply prow/config/config.go]: That is definitely simpler but we are totally inconsistent as to when it does or does not allow for an override, which the tiered way does. #9757 

/shrug

Comment (cjwagner) [reply prow/config/config.go]: /shrug

Comment (alvaroaleman) [reply prow/config/config.go]: True. Ironically I've thought about doing it like that initially but ditched the idea because there is no safe way to distinguish repos from orgs. But yeah, enforcing to prefix repos with `org/` is the obvious solution for that.

I've changed it, IMHO makes both the resulting config and the tests much easier to read. LMK what you think.

File: prow/config/config.go (hunk 7 above)
Comment (alvaroaleman) [reply prow/config/config.go]: Updated

File: prow/config/config.go [hunk 9]
```diff
@@ -174,9 +174,27 @@ type Plank struct {
...
 	// JobURLPrefix is the host and path prefix under
 	// which job details will be viewable
+	// TODO @alvaroaleman: Remove in September 2019
 	JobURLPrefix string `json:"job_url_prefix,omitempty"`
+	// JobURLPrefixConfig is the host and path prefix under which job details
+	// will be viewable. Use `org/repo`, `org` or `*`as key and an url as value
+	JobURLPrefixConfig map[string]string `json:"job_url_prefix_config,omitempty"`
+}
+
+func (p Plank) GetJobURLPrefix(refs *prowapi.Refs) string {
```
Comment (krzyzacy) [inline prow/config/config.go]: eh...... it's this a breaking change? We should still fall back to `JobURLPrefix` if `JobURLPrefixConfig` is empty

Comment (krzyzacy) [reply prow/config/config.go]: https://github.com/kubernetes/test-infra/blob/master/prow/config.yaml#L4, /shrug

cc @srinivashegde86
